django/core/signing.py

```diff
@@ -266,12 +267,28 @@ def unsign(self, value, max_age=None):
         """
         result = super().unsign(value)
         value, timestamp = result.rsplit(self.sep, 1)
-        timestamp = b62_decode(timestamp)
-        if max_age is not None:
-            if isinstance(max_age, datetime.timedelta):
-                max_age = max_age.total_seconds()
-            # Check timestamp is not older than max_age
-            age = time.time() - timestamp
-            if age > max_age:
-                raise SignatureExpired("Signature age %s > %s seconds" % (age, max_age))
+        self.payload_timestamp = b62_decode(timestamp)
+        self._verify_max_age(max_age)
+        return value
+
+    def unsign_object(self, signed_obj, max_age=None, expiration_key=None, **kwargs):
+        # First verify the outer max_age boundary
+        value = super().unsign_object(signed_obj, max_age=max_age, **kwargs)
+
+        # Then if expiration_key is provided, check it as an additional constraint
+        if expiration_key is not None:
+            expiry = value.get(expiration_key)
```

Comment:
`value` here is allowed to be a "complex data structure (e.g. list, tuple, or dictionary)". Should support be added for non-dictionary types (perhaps supporting indexing?). Alternatively, should the error message be nicer if it is of the wrong type?

Reply:
Thanks for raising this point! You're right that we should handle non-dictionary types more gracefully.

Here's what I propose adding:

```python
if not isinstance(value, dict):
    raise TypeError(
        f"expiration_key is only supported for dict values, not {type(value).__name__}"
    )

Reply:
is there is a more abstract key indexable type we can check for?
this way we can support more than just a `dict`

Reply:
:+1: Nice catch! I hadn't thought that far to support more than just a regular `dict`. I think we can use [Mapping](https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping). 

We can also add tests for non-supported type input.